### PR TITLE
[RLlib] Bring back BC and Marwil learning tests.

### PR DESCRIPTION
**Pull request by gjoliver**
Also make sure we can stop trials early with evaluation metrics.

## Why are these changes needed?

Now that the bc and marwil are verifeid, we should start testing them.
Note that some minor fixes had to be made before these trials can be properly terminated based on their eval results.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [*] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/agents/trainer.py [hunk 1]
```diff
@@ -723,7 +723,17 @@ def default_logger_creator(config):
...
-        self.evaluation_metrics = {}
+        # Initialize common evaluation_metrics to nan, before they become
+        # available. We want to make sure the metrics are always present
+        # (although their values may be nan), so that Tune does not complain
+        # when we use these as stopping criteria.
+        self.evaluation_metrics = {
+            "evaluation": {
+                'episode_reward_max': np.nan,
+                'episode_reward_min': np.nan,
+                'episode_reward_mean': np.nan,
```
Comment (sven1977) [inline rllib/agents/trainer.py:734]:
```suggestion
                "episode_reward_max": np.nan,
                "episode_reward_min": np.nan,
                "episode_reward_mean": np.nan,
```

**Review by sven1977 — APPROVED** [review]
Nice! Thanks for digging into all of this and reviving these important tests, @gjoliver ! :)

**Comment by sven1977** [comment]
Re-merged from upstream master (some CI stuff was broken). Waiting for tests to pass.
cc: @gjoliver

**Comment by sven1977** [comment]
Fixed test_trainer.py. One more try :) ...
I took out the `hist_stats: {}` again. I don't think it makes sense to leave this arbitrarily defined just because of that one test case. Fixed the test instead. Also, now, even after the very first iteration, there is a "evaluation" result, so I changed the test from `assertFalse(r0 ...)` to `assertTrue(r0 ...)`.

...

**Comment by gjoliver** [comment]
> 

thanks. yeah, I did the same to that test. probably missed the assertFalse check. thanks.

**Pull request merged**